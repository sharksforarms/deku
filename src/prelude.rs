pub use crate::{error::DekuError, BitsReader, BitsWriter, DekuRead, DekuWrite};
pub use bitvec::{
    order::BitOrder, order::Lsb0, order::Msb0, slice::AsBits, slice::BitSlice, vec::BitVec,
};
