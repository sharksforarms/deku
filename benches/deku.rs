use criterion::{black_box, criterion_group, criterion_main, Criterion};
use deku::prelude::*;

#[derive(Debug, PartialEq, DekuRead)]
struct DekuByte {
    data: u8
}

#[derive(Debug, PartialEq, DekuRead)]
#[deku(id_type = "u8")]
enum DekuEnum {
    #[deku(id = "0x01")]
    VariantA(u8),
}

#[derive(Debug, PartialEq, DekuRead)]
struct DekuVec {
    count: u8,
    #[deku(len = "count")]
    data: Vec<u8>
}

fn deku_read_byte(input: &[u8]) {
    let (_rest, v) = DekuByte::from_bytes((input, 0)).unwrap();
    assert_eq!(DekuByte { data: 0x01 }, v);
}

fn deku_read_enum(input: &[u8]) {
    let (_rest, v) = DekuEnum::from_bytes((input, 0)).unwrap();
    assert_eq!(DekuEnum::VariantA(0x02), v);
}

fn deku_read_vec(input: &[u8]) {
    let (_rest, v) = DekuVec::from_bytes((input, 0)).unwrap();
    assert_eq!(100, v.count);
    assert_eq!(100, v.data.len());
}

fn criterion_benchmark(c: &mut Criterion) {
    c.bench_function("deku_read_byte", |b| {
        b.iter(|| deku_read_byte(black_box([0x01].as_ref())))
    });

    c.bench_function("deku_read_enum", |b| {
        b.iter(|| deku_read_enum(black_box([0x01, 0x02].as_ref())))
    });

    let deku_read_vec_input = {
        let mut v = [0xFFu8; 101].to_vec();
        v[0] = 100u8;
        v
    };
    c.bench_function("deku_read_vec", |b| {
        b.iter(|| deku_read_vec(black_box(&deku_read_vec_input)))
    });
}

criterion_group!(benches, criterion_benchmark);
criterion_main!(benches);
