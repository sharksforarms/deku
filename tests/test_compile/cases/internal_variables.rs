use deku::prelude::*;

#[derive(DekuRead, DekuWrite)]
struct TestCount {
    field_a: u8,
    #[deku(count = "deku::byte_offset")]
    field_b: Vec<u8>
}

#[derive(DekuRead, DekuWrite)]
struct TestBitRead {
    field_a: u8,
    #[deku(bits_read = "deku::bit_offset")]
    field_b: Vec<u8>
}

#[derive(DekuRead, DekuWrite)]
struct TestBytesRead {
    field_a: u8,
    #[deku(bytes_read = "deku::bit_offset")]
    field_b: Vec<u8>
}

#[derive(DekuRead, DekuWrite)]
struct TestUntil {
    field_a: u8,
    #[deku(until = "|v| *v as usize == deku::bit_offset")]
    field_b: Vec<u8>
}

#[derive(DekuRead, DekuWrite)]
struct TestCond {
    field_a: u8,
    #[deku(cond = "deku::bit_offset == *field_a as usize")]
    field_b: u8
}

#[derive(DekuRead, DekuWrite)]
struct TestDefault {
    field_a: u8,
    #[deku(skip, default = "deku::byte_offset")]
    field_b: usize
}

#[derive(DekuRead, DekuWrite)]
struct TestMap {
    field_a: u8,
    #[deku(map = "|v: u8| -> Result<_, DekuError> { Ok(v as usize + deku::byte_offset) }")]
    field_b: usize
}

fn dummy_reader(
    offset: usize,
    rest: &BitSlice<Msb0, u8>,
) -> Result<(&BitSlice<Msb0, u8>, usize), DekuError> {
    Ok((rest, offset))
}
#[derive(DekuRead, DekuWrite)]
struct TestReader {
    field_a: u8,
    #[deku(reader = "dummy_reader(deku::byte_offset, deku::rest)")]
    field_b: usize
}

#[derive(DekuRead, DekuWrite)]
#[deku(ctx = "_byte_size: usize, _bit_size: usize")]
struct ChildCtx {
}
#[derive(DekuRead, DekuWrite)]
struct TestCtx {
    field_a: u8,
    #[deku(ctx = "deku::byte_offset, deku::bit_offset")]
    field_b: ChildCtx
}

fn dummy_writer(
    _offset: usize,
    _output: &mut BitVec<Msb0, u8>,
) -> Result<(), DekuError> {
    Ok(())
}
#[derive(DekuRead, DekuWrite)]
struct TestWriter {
    field_a: u8,
    #[deku(writer = "dummy_writer(deku::byte_offset, deku::output)")]
    field_b: usize
}

#[derive(DekuRead, DekuWrite)]
struct FailInternal {
    field_a: u8,
    #[deku(cond = "__deku_bit_offset == *field_a as usize")]
    field_b: u8
}

fn main() {}
