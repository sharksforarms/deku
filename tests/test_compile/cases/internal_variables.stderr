error: proc-macro derive panicked
  --> $DIR/internal_variables.rs:89:20
   |
89 | #[derive(DekuRead, DekuWrite)]
   |                    ^^^^^^^^^
   |
   = help: message: error: attribute cannot contain `__deku_` these are internal variables. Please use the `deku::` instead.

error: proc-macro derive panicked
  --> $DIR/internal_variables.rs:89:10
   |
89 | #[derive(DekuRead, DekuWrite)]
   |          ^^^^^^^^
   |
   = help: message: error: attribute cannot contain `__deku_` these are internal variables. Please use the `deku::` instead.
